vpat 44: scaffold keyboard tab selection focus

abaevbog opened the pull request:
Do not move focus from the tab onto the editor/input, during keyboard navigation to not change context per https://www.w3.org/WAI/WCAG21/Understanding/on-input. 

Based on the success criteria above, it seems like it is desired to avoid context (specifically focus in our case) changes during tab navigation in general. Avoiding it completely would be detrimental to overall functionality, since it's way better for the input to be focused after you click on a tab. So I thought this could be a reasonable compromise that's also somewhat close to what firefox does: if you switch tabs via clicking, the focus will move into the selected tab panel but if you select a tab with keyboard, focus remains on it?

---

VPAT 44 description for reference:

In the Zotero desktop app's 'Translator'/'Scaffold' modal, there is a tab panel with tabs for different sections. The 'Code,' Tests,' and 'Test Input' tabs have code inputs in them. When one of these tabs is selected, the code input is automatically focused. Focus should remain on the tab panel. This occurs wiht or without a screen reader active.

Impact Statement: Operable elements have certain predictable behaviors. Whenever something breaks those expectations, the user should be informed ahead of time. If they are not it can be disorienting for all users, but especially those with visual or cognitive impairments.

Expected Result: Changing the setting of an element does not cause an unexpected change of context.

abaevbog commented on chrome/content/scaffold/scaffold.js:
```diff
@@ -871,8 +879,16 @@ var Scaffold = new function () {
 					_editors[tab].focus();
 				}, 10);
 				break;
+			default:
```
What about the first tab, `metadata`? I was under the impression that both `metadata` and `browser` have their first focusable element focused after selection by default, and this kinda does just that, except after a small delay. I could break this up into two separate cases - would it be more readable that way?

For context, I don't think this part is strictly required - the main point with these changes is to not have the focus go away from tabs during keyboard navigation. If you switch to voiceover, click around, and then turn it off, the focus may not start going back to the fields after one or two tab selections. So this is just for consistency rather than anything else

dstillman commented:
Not sure if it's from this, but clicking on the Tests tab should focus the test list, not the code editor.

dstillman commented:
Seems like there's a keyboard trap in the editor of the Code tab. Once the code editor has focus, I can't get out of it. Shift-Tab from the beginning of the editor should move focus to the tab bar. Same in Test Input.

(We're also not showing a focus ring on the tab bar, but that's not a VPAT issue, so we can deal with that separately.)

dstillman commented:
Maybe Esc in the editor should also move focus back to the tab bar?

dstillman commented:
Oh, though you can still use Ctrl-Tab to move to the next tab even when you're in the editor. But should probably still provide another way to get out of it.

abaevbog commented:
> Not sure if it's from this, but clicking on the Tests tab should focus the test list, not the code editor.

Yeah, this is the current behavior. I changed it to place focus  on the test list.

> Seems like there's a keyboard trap in the editor of the Code tab. Once the code editor has focus, I can't get out of it. Shift-Tab from the beginning of the editor should move focus to the tab bar. Same in Tests and Test Input.

Gotcha, Shift-Tab from the beginning of a line in an editor will focus the previous element out of the editor. It is not quite standard (e.g. vscode does not do that and you move focus between sections with shortcuts) but it does feel quite natural to me here.

> (We're also not showing a focus ring on the tab bar, but that's not a VPAT issue, so we can deal with that separately.)

I added the outline with a few small tweaks for it to not get cutoff and to not flash on mouse click.

> Maybe Esc in the editor should also move focus back to the tab bar, so there's a way to do it without being at the beginning of the content?

That makes sense! It doesn't seem like Escape is being used for anything else, so I added it too.

dstillman commented:
> Shift-Tab from the beginning of a line in an editor will focus the previous element out of the editor

Sorry, I was just saying the very beginning of the full content block, not the beginning of each line.

abaevbog commented:
Ah, gotcha! This will now tab out on shift-tab from the beginning of the first line of the editor only.

I also changed to use monaco API to get the position of the cursor, which is way cleaner than checking the iframe's DOM as before.

abaevbog commented:
That is a great point! Yeah, we should not move focus away if the user wants to just dismiss the autocomplete popup, for sure.

When I tried monaco's `onKeyDown`, it seems to behave somewhat similar to a capturing listener emitting absolutely all events, regardless if they are handled somehow by the editor or not...

But if we just pull our Escape handling into a separate non-capturing listener (shift-tab still has to be captured), I think we get what we want. If there is an autocomplete popup, the first Escape will not reach our listener, so it will just hide the popup. The second Escape, however, will bubble and we will move focus to the tabbar. What do you think?

abaevbog commented:
That makes sense! Since there are two listeners (capturing and non-capturing) meant for a related purpose, I thought we should keep them together. So I added `addEditorKeydownHandlers(editor)` called with `_editors.import`, `_editors.tests` and `_editors.code` as parameters. In that function, we add the listeners directly to the document of the editors (`editor.getDomNode().ownerDocument;`). How does that sound?

dstillman commented:
Fine with me.

abaevbog commented:
If it is interfering with existing workflows, there is no reason to keep it, since one still has a way to get out of the editor without having to use a mouse (either on shift-tab from the start of the editor or via Command-{number}). So yeah, we can remove it!
